[REEF-126]: Merge Enhanced Features into Shimoga

jsjason opened the pull request:
This addressed the issue by
  * adding Gather and Scatter operations and the corresponding configurations
  * modifying the OperatorTopology classes to support Gather and Scatter
  * changing the TopologyUpdate message so that
    the Driver sends list of currently active Tasks for Scatter

JIRA: [REEF-126](https://issues.apache.org/jira/browse/REEF-126)

bgchun commented:
@shravanmn Could you review this pull request? This issue has been long over due. It'd be great to wrap it up. Thanks!

jsjason commented:
I tested the operations in a similar fashion as Broadcast/Reduce in the BroadcastREEF example, although I didn't include it in this PR.

markusweimer commented:
>I tested the operations in a similar fashion as Broadcast/Reduce in the BroadcastREEF example, although I didn't include it in this PR.

Any reason not to include the tests? The more tests, the better :-)

jsjason commented:
I used the term test, but actually I just hard-coded the BroadcastREEF example to include Scatter and Gather operations. They're not appropriate as a REEF example, I think. We sure could use some unit tests, though. (https://issues.apache.org/jira/browse/REEF-187)

shravanmn commented:
Will take a look.

bgchun commented:
@shravanmn Can you take a look? If you cannot, I will take a look and merge this. 

@jsjason Can you address merge conflicts? Let's review this request in time.

bgchun commented:
@jsjason We haven't heard from @shravanmn yet. Can you address the conflicts so that I can review your code? Thanks.

bgchun commented:
Copying Markus's comment. "The build servers merge current master into
the code before building. And we recently made rat part of the build.
If you merge (or rebase onto) master on your side, you should see the
same behavior / error."


[INFO] ------------------------------------------------------------------------Waiting
for Jenkins to finish collecting data[ERROR] Failed to execute goal
org.apache.rat:apache-rat-plugin:0.11:check (validate) on project
reef-dotnet: Too many files with unapproved license: 2 See RAT report
in: /home/jenkins/jenkins-slave/workspace/Reef-pull-request-ubuntu/lang/cs/target/rat.txt
-> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors,
re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X
switch to enable full debug logging.[ERROR] [ERROR] For more
information about the errors and possible solutions, please read the
following articles:[ERROR] [Help 1]
http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException[ERROR]
[ERROR] After correcting the problems, you can resume the build with
the command[ERROR]   mvn <goals> -rf :reef-dotnet


On Sun, Jul 5, 2015 at 5:15 PM, asfbot <notifications@github.com> wrote:

> Reef-pull-request-ubuntu #6
> <https://builds.apache.org/job/Reef-pull-request-ubuntu/6/> FAILURE
> Looks like there's a problem with this pull request
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/226#issuecomment-118594851>
> .
>



-- 
Byung-Gon Chun

jsjason commented:
This is a bit strange. I ran all tests and `mvn apache-rat:check` after deleting my .m2 folder, but no errors occur for me, both on Mac and Ubuntu. The error message that the build server generates:
```
...
[INFO] Exclude: Org.Apache.REEF.Common/Avro/*
[INFO] Exclude: Org.Apache.REEF.Tests/ConfigFiles/evaluator.conf
[INFO] Exclude: Org.Apache.REEF.Tang.Tests/evaluator.conf
[INFO] Exclude: Org.Apache.REEF.Tang.Tests/simpleConstructorJavaProto.bin
[INFO] 838 resources included (use -debug for more details)
[INFO] Rat check: Summary of files. Unapproved: 2 unknown: 2 generated: 0 approved: 833 licence.
...
```
appears as 
```
...
[INFO] Exclude: Org.Apache.REEF.Common/Avro/*
[INFO] Exclude: Org.Apache.REEF.Tests/ConfigFiles/evaluator.conf
[INFO] Exclude: Org.Apache.REEF.Tang.Tests/evaluator.conf
[INFO] Exclude: Org.Apache.REEF.Tang.Tests/simpleConstructorJavaProto.bin
[INFO] 15 resources included (use -debug for more details)
Warning:  org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser: Property 'http://www.oracle.com/xml/jaxp/properties/entityExpansionLimit' is not recognized.
Compiler warnings:
  WARNING:  'org.apache.xerces.jaxp.SAXParserImpl: Property 'http://javax.xml.XMLConstants/property/accessExternalDTD' is not recognized.'
Warning:  org.apache.xerces.parsers.SAXParser: Feature 'http://javax.xml.XMLConstants/feature/secure-processing' is not recognized.
Warning:  org.apache.xerces.parsers.SAXParser: Property 'http://javax.xml.XMLConstants/property/accessExternalDTD' is not recognized.
Warning:  org.apache.xerces.parsers.SAXParser: Property 'http://www.oracle.com/xml/jaxp/properties/entityExpansionLimit' is not recognized.
[INFO] Rat check: Summary of files. Unapproved: 0 unknown: 0 generated: 0 approved: 15 licence.
```
to me. It seems that the build server is recognizing something that my local machine isn't.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/config/GatherOperatorSpec.java:
```diff
@@ -0,0 +1,84 @@
+
+  public GatherOperatorSpec(final String receiverId,
+                            final Class<? extends Codec> dataCodecClass) {
+    super();
```
super() ?

jsjason replied:
I added `super()` to the constructor of this class and `ScatterOperatorSpec` because the other `OperatorSpec` classes did too, but since these classes are not extending another concrete class or abstract class, I guess we can just take it out. I'll remove it.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/config/GatherOperatorSpec.java:
```diff
@@ -0,0 +1,84 @@
+
+  @Override
+  public String toString() {
+    return "Gather Operator Spec: [receiver=" + receiverId + "] [dataCodecClass=" + Utils.simpleName(dataCodecClass)
```
Use StringBuilder.

jsjason replied:
Modified `GatherOperatorSpec.toString()` and `ScatterOperatorSpec.toString()` to use `StringBuilder`.

bgchun commented:
@jsjason Overall, it looks fine. Did you address Shravan's comments? Can you update the commit message to reflect any limitation if there is? 

There is no test. Can we add tests soon, which is another JIRA issue? Maybe a good intern exercise?

bgchun commented:
@jsjason In addition to basic scatter and gather tests, we need to add tests that invoke failure handling.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/operators/GatherReceiver.java:
```diff
@@ -0,0 +1,171 @@
+  private final NetworkService<GroupCommunicationMessage> netService;
+  private final Sender sender;
+
+  private final OperatorTopology topology;
```
Line space between member variables. Use a consistent style.

jsjason replied:
Got it. I'll list all fields without any empty lines, expect the only static field `LOG`.

shravanmn commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/task/CommunicationGroupClientImpl.java:
```diff
@@ -191,6 +248,8 @@ public void initialize() {
         throw new RuntimeException("InterruptedException while waiting for initialization", e);
       }
 
+      updateTopology();
```
updateTopology? IIRC this is to be called only by master task from user code. I think you are trying to use this to bootstrap active tasks. I am not sure if this is the right way. initialize() will be called by all the tasks.

jsjason replied:
You're right. I included this statement to collect the list of active tasks during initialization, but forgot  the fact that only the `ScatterSender` needs to do this. I'll refactor this class so that only `ScatterSenders` perform this update during initialization.

shravanmn commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/task/OperatorTopologyStructImpl.java:
```diff
@@ -325,6 +326,49 @@ public void sendToChildren(final byte[] data, final ReefNetworkGroupCommProtos.G
   }
 
+  @Override
+  public byte[] recvFromChildren() {
```
I see a lot of code duplication. Can you refactor instead?

jsjason replied:
Sure. I'll extract the `EMPTY_BYTE_ARR` checking part as a separate function and make `recvFromParent()` and `recvFromChildren()` use that function.

shravanmn commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/operators/GatherReceiver.java:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  @Override
+  public List<T> receive() throws NetworkException, InterruptedException {
```
I believe that this interface guarantees that the data received will be in lexicographically sorted id order. But I don't see this anywhere in the code.

jsjason replied:
Sorry about that, there weren't any comments in the `Gather` interface that mentions lexicographical order requirements. I can refactor the code to return elements in lexicographically sorted id order, but before I do I'd like to politely ask why we'd want to implement `Gather` that way. In an elastic environment, the task identifiers wouldn't matter much, would they? Please correct me if I'm wrong.

shravanmn commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/operators/ScatterReceiver.java:
```diff
@@ -0,0 +1,162 @@
+      final byte[] data = topology.recvFromParent(ReefNetworkGroupCommProtos.GroupCommMessage.Type.Scatter);
+
+      LOG.fine(this + " Trying to send scatter to children");
+      topology.sendToChildren(data, ReefNetworkGroupCommProtos.GroupCommMessage.Type.Scatter);
```
I am a little confused. I thought we got the active tasks and all that so that we do not have to send the whole data to the children. Am I missing something?

jsjason replied:
Sorry, I think I'm the one who misunderstood our conversation on JIRA. I'll change the code so that  we use a encoded topology to send only relevant data.

shravanmn commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/FlatTopology.java:
```diff
@@ -285,6 +285,15 @@ private void processUpdateTopology(final GroupCommunicationMessage msg) {
     nodeTopologyUpdateWaitStage.onNext(toBeUpdatedNodes);
   }
 
+  private byte[][] getByteArrayOfActiveSlaveTasks() {
```
I thought this would encode topology and not just the currently active tasks.

jsjason replied:
I misunderstood our conversation on JIRA. I'll change this.

shravanmn commented:
I took a look and left some comments. I thought this pull request is implementing the strategy we discussed in [REEF-126](https://issues.apache.org/jira/browse/REEF-126). But I don't see that. Please let me know if I am missing something.

jsjason commented:
I'll push another commit based on @bgchun and @shravanmn's comments. While I'm at it, I'd appreciate it if we talk more about the `Gather` lexicographical order issue comment @shravanmn left.

bgchun commented:
known MPI gather semantics: the elements are ordered by the rank of the
process from which they were received.

@shravanmn, are you talking about the elements are ordered by the task id
lexicographically?


On Thu, Jul 9, 2015 at 9:52 PM, Jason (Joo Seong) Jeong <
notifications@github.com> wrote:

> I'll push another commit based on @bgchun <https://github.com/bgchun> and
> @shravanmn <https://github.com/shravanmn>'s comments. While I'm at it,
> I'd appreciate it if we talk more about the Gather lexicographical order
> issue comment @shravanmn <https://github.com/shravanmn> left.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/226#issuecomment-119952056>
> .
>



-- 
Byung-Gon Chun

shravanmn commented:
Yes, since task ids are the closest to the ranks, when we initially designed the API, we decided that when the user does not specify an ordering, we will assume the lexicographic ordering of task-ids as the default order.

jsjason commented:
Okay, I understand now. I'll change the implementation so that `GatherReceive.receive()` returns object sorted in the lexicographical ordering of task ids.

jsjason commented:
@shravanmn and @bgchun, I've addressed your comments in the push I just made. Could you take another look?
* Our discussion on the Scatter API on [REEF-126](https://issues.apache.org/jira/browse/REEF-126) ended without a conclusion, so I left all methods in (all implemented) for now.
* Lexicographical order on task ids is taken when no other order is specified, both for Scatter and Gather. 
* I also included a couple of tests for the helper classes, but didn't write any other tests since we have [REEF-187](https://issues.apache.org/jira/browse/REEF-187). Nonetheless, I have confirmed that all Gather APIs and Scatter APIs function correctly by slightly modifying the BroadcastREEF example, not included here.

bgchun commented:
@jsjason You addressed my comments.
@shravanmn Could you check if Jason addressed your comments? Can we also resolve the Scatter API?

This issue currently blocks [REEF-420](https://issues.apache.org/jira/browse/REEF-420).

dafrista commented:
Pinging @jsjason @bgchun @shravanmn as release manager of 0.12. This issue is considered a critical "blocking" issue.

From a quick look at the discussion, it looks like your move @shravanmn :)

jsjason commented:
I pushed another commit that resolves checkstyle issues. Ready for review anytime.

markusweimer commented:
@shravanmn I believe this needs your review. Right after @jsjason has resolved the merge conflicts :-)

jsjason commented:
Thanks, @markusweimer. I wouldn't have noticed there were conflicts again if you hadn't told me. :+1: I pushed yet another commit that fixes the conflicts.

shravanmn commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/operators/ScatterReceiver.java:
```diff
@@ -0,0 +1,164 @@
+      }
+
+      LOG.fine(this + " Successfully received scattered data.");
+      final Pair<byte[][], Map<String, byte[]>> myDataAndChildBytesMap = ScatterHelper.decodeScatterMsgForNode(data);
```
Feels like this should be a separate class. I think it will also simplify ScatterHelper part. I would have this object and a codec for it instead of the helper.

jsjason replied:
I agree. I'll wrap `Pair<byte[][], Map<String, byte[]>>` as a class and define a decoder for it.

shravanmn commented:
Apart from one comment looks good to me. If you want to you can address it or take it up later. It is not a major issue but would improve quality of code I guess. Perhaps a few unit tests or an application as an end to end test might be helpful. I won't be able to merge this but code looks good to me.

jsjason commented:
I've addressed @shravanmn's comments. This is ready for another review (or merge).

dafrista commented on lang/java/reef-io/src/test/java/org/apache/reef/io/network/group/impl/utils/ScatterHelperTest.java:
```diff
@@ -0,0 +1,118 @@
+   * Create a small topology of 4 nodes and simulate a scatter operation.
+   */
+  @Test
+  public void testEncodeDecode() {
```
This test seems out of place here. Would a new test class named `ScatterCodec` make more sense?

jsjason replied:
Good point. `ScatterHelperTest` should only test methods of `ScatterHelper`. I'll move this test to a new  class named `ScatterCodecTest`.

dafrista commented on lang/java/reef-io/src/test/java/org/apache/reef/io/network/group/impl/utils/ScatterHelperTest.java:
```diff
@@ -0,0 +1,118 @@
+    final Map<String, byte[]> encodedDataMap = scatterEncoder.encode(elements, counts, taskOrder, codec);
+
+    // check msg correctness for childNode1 (Task-1)
+    ScatterData scatterData = scatterDecoder.decode(encodedDataMap.get(childNode1.getTaskId()));
```
The re-assignment of `ScatterData` seems to go against our conventions. How about splitting up into `final ScatterData childNode1Data` and `final ScatterData childNode2Data`?

dafrista commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/utils/ScatterData.java:
```diff
@@ -0,0 +1,50 @@
+public final class ScatterData {
+
+  /**
+   * Data that is assigned to this node.
```
If we move these comments to the public getters, they will be available in javadoc.

dafrista commented:
@jsjason I reviewed some of the newest code. I'll test and merge once these are addressed.

dafrista commented:
Great, I'll do the merge. Good work @jsjason @shravanmn @bgchun @markusweimer

jsjason commented:
Thanks for the reviews, @shravanmn, @bgchun, @markusweimer, and @dafrista.
